### PR TITLE
Implement HSM and Edgelet API to obtain certificate common name

### PR DESCRIPTION
**Pull request by mrohera**
- HSM API to retrieve common name was defined but not implemented. Changes include parsing of the CN field of the certificate subject. Unit and integration tests have also been added.

- API to retrieve the common name from Edgelet core - HSM was implemented along with supporting tests.

### CODE REVIEW

File: edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c [hunk 1]
```diff
@@ -92,6 +91,23 @@ static const int month_day[] = { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 30
...
+// https://tools.ietf.org/html/rfc5280#appendix-A
+#define MAX_LEN_COMMON_NAME 64
+// Common name OID 2.5.4.3
+#define COMMON_NAME_OID_SIZE 5
+static const unsigned char COMMON_NAME_OID[COMMON_NAME_OID_SIZE] = {  0x06, 0x03, 0x55, 0x04, 0x03 };
+
+#define FREEIF(x) \
+    do { \
+        if ((x != NULL)) { \
+            free(x); \
```
Comment (avranju) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c]: Isn't passing `NULL` for `free` allowed? From the standard:
> **7.22.3.3**
> The `free` function causes the space pointed to by `ptr` to be deallocated, that is, made available for further allocation.  If `ptr` is a null pointer, no action occurs.

File: edgelet/edgelet-hsm/src/crypto.rs [hunk 2]
```diff
@@ -183,4 +183,11 @@ impl CoreCertificate for Certificate {
             .map_err(|err| Error::from(err.context(ErrorKind::Hsm)))
             .map_err(|err| CoreError::from(err.context(CoreErrorKind::KeyStore)))
     }
+
+    fn get_common_name(&self) -> Result<String, CoreError> {
+        self.0
+            .get_common_name()
+            .map_err(|err| Error::from(err.context(ErrorKind::Hsm)))
+            .map_err(|err| CoreError::from(err.context(CoreErrorKind::KeyStore)))
```
Comment (avranju) [inline edgelet/edgelet-hsm/src/crypto.rs]: I see that you're following the pattern from the other existing methods, but is somewhat curious that we are using 2 `map_err` calls to do this transformation. Would having only the second `map_err` be acceptable? Just that you wouldn't have the `ErrorKind::Hsm` error in the error stack.

File: edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c [hunk 3]
```diff
@@ -441,6 +464,82 @@ static int parse_tbs_cert_info(unsigned char* tbs_info, size_t len, CERT_DATA_IN
...
+        }
+        else
+        {
+            ASN1_OBJECT seq_obj;
+            if (memcmp(COMMON_NAME_OID, oid_start, COMMON_NAME_OID_SIZE) == 0)
+            {
+                // back up 2 bytes to ensure we are at a ASN1_SEQUENCE boundary
+                // note it will be 2 bytes because the CN is 64 characters which
+                // and its oid is 5 chars which and their sum is less than 127
+                // and thus will have length expressed in 1 byte
```
Comment (avranju) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c]: Nit: Redundant `which` in comment.

**Comment by mrohera** [comment]
A quick comment about the C vs Rust, I would like this to be in Rust and perhaps there is way to make this happen. We can chat about this offline. Also I am thinking of not using this and rely on openssl directly and not use this parsing code. This however could be used in the C HSM client which does not necessarily rely on openssl.

File: edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c (hunk 1 above)
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c]: Cool good to know but I don't feel like changing this.

File: edgelet/edgelet-hsm/src/crypto.rs (hunk 2 above)
Comment (mrohera) [reply edgelet/edgelet-hsm/src/crypto.rs]: I think I prefer the HSM in the stack to quickly narrow down the scope of the issue. I did however amend the ErrorKind to be a bit more accurate.

**Pull request merged**